presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.symbolAllocator = symbolAllocator;
+            this.session = session;
+            this.lookup = lookup;
+            this.resultOrdering = getResultOrdering(costComparator, session);
+            this.allInference = createEqualityInference(filter);
+            this.allFilter = filter;
+        }
+
+        private static Ordering<JoinEnumerationResult> getResultOrdering(CostComparator costComparator, Session session)
+        {
+            return new Ordering<JoinEnumerationResult>()
+            {
+                @Override
+                public int compare(JoinEnumerationResult result1, JoinEnumerationResult result2)
+                {
+                    checkArgument(result1.cost.isPresent() && result2.cost.isPresent(), "costs not present");
+                    return costComparator.compare(session, result1.cost.get(), result2.cost.get());
+                }
+            };
+        }
+
+        private JoinEnumerationResult chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            JoinEnumerationResult bestResult = memo.get(multiJoinKey);
+            if (bestResult == null) {
+                checkState(sources.size() > 1);
+                List<JoinEnumerationResult> results = generatePartitions(sources.size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
+                        .collect(toImmutableList());
+                if (results.stream().anyMatch(result -> result.cost.isPresent() && result.cost.get().isUnknown())) {
+                    return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+                }
+                bestResult = results.stream()
+                        .filter(result -> result.planNode.isPresent())
+                        .min(resultOrdering)
+                        .orElseGet(() -> createCrossJoins(sources, outputSymbols));
+
+                // we need to check for empty cost again in case we created a cross join with unknown cost
+                if (bestResult.cost.get().isUnknown()) {
+                    return new JoinEnumerationResult(Optional.of(UNKNOWN_COST), Optional.empty());
+                }
+                log.debug("Least cost join was: " + bestResult.planNode.get().toString());
+                memo.put(multiJoinKey, bestResult);
+            }
+            return bestResult;
+        }
+
+        /**
+         * This method produces a cross join of all the sources in multiJoinNode.
+         * This gets called once there are no longer edges between any of the join sources.
+         */
+        private JoinEnumerationResult createCrossJoins(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            return sources.stream()
```

Comment:
let's assume we have sources: `s1, s2, s3, ... sn`. Does `createCrossJoins` create join:
```
s1 join s2 join s3 join s4 ... join sn
```
? In https://github.com/Teradata/presto/pull/584#discussion_r123779783 you mentioned that `createCrossJoins does costing though and determines which table goes to which side. Leaving as is.` But since the order of cross join is predefined than I think it can be worse than original plan, can't it?